### ISSUE TITLE
Reproduce the training of ComplEx-N3 in CQD

### ISSUE THREAD

**Issue by KiddoZhu** [pminervini/KGReasoning#3]
Hi @pminervini! Thanks for this awesome implementation of CQD.

@migalkin and I are trying to reproduce the results of CQD. For the inference part mentioned in https://github.com/uclnlp/cqd/issues/7, we found that using product t-norm for all query types is enough to get good numbers. The important thing is to set `--cqd-sigmoid`, which is not turned on by default in the argparser. Maybe it's better to set this argument to true by default?

However, we met some troubles when reproducing the training of ComplEx-N3 in CQD. If we understand correctly, the training is carried out with the same codebase, except using only 1p task and the QA iterator. We exactly follow the hyperparameters in `config.json` in the checkpoint directory, and here are the beginning of our training log on FB15k-betae.

```
2021-12-29 12:36:19 INFO     Training average positive_sample_loss at step 0: 19.225096
2021-12-29 12:36:19 INFO     Training average negative_sample_loss at step 0: 19.225096
2021-12-29 12:36:19 INFO     Training average loss at step 0: 19.225096
...
2021-12-29 12:47:15 INFO     Training average positive_sample_loss at step 1900: 18.688958
2021-12-29 12:47:15 INFO     Training average negative_sample_loss at step 1900: 18.688958
2021-12-29 12:47:15 INFO     Training average loss at step 1900: 18.688958
2021-12-29 12:47:50 INFO     Evaluating on Valid Dataset...
2021-12-29 12:48:35 INFO     Valid 1p MRR at step 2000: 0.005592
2021-12-29 12:48:35 INFO     Valid 1p HITS1 at step 2000: 0.000137
2021-12-29 12:48:35 INFO     Valid 1p HITS3 at step 2000: 0.000373
2021-12-29 12:48:35 INFO     Valid 1p HITS10 at step 2000: 0.011451
2021-12-29 12:48:35 INFO     Valid 1p num_queries at step 2000: 59078.000000
2021-12-29 12:48:35 INFO     Valid average MRR at step 2000: 0.005592
2021-12-29 12:48:35 INFO     Valid average HITS1 at step 2000: 0.000137
2021-12-29 12:48:35 INFO     Valid average HITS3 at step 2000: 0.000373
2021-12-29 12:48:35 INFO     Valid average HITS10 at step 2000: 0.011451
```

This is very different from the loss and MRR in the training log of the official checkpoint.

```
2021-06-07 06:40:01,518 INFO     Training average positive_sample_loss at step 0: 19.225096
2021-06-07 06:40:01,518 INFO     Training average negative_sample_loss at step 0: 19.225096
2021-06-07 06:40:01,518 INFO     Training average loss at step 0: 19.225096
2021-06-07 06:45:01,820 INFO     Evaluating on Valid Dataset...
2021-06-07 06:45:49,154 INFO     Valid 1p MRR at step 500: 0.777703
2021-06-07 06:45:49,154 INFO     Valid 1p HITS1 at step 500: 0.679669
2021-06-07 06:45:49,154 INFO     Valid 1p HITS3 at step 500: 0.860898
2021-06-07 06:45:49,154 INFO     Valid 1p HITS10 at step 500: 0.931428
2021-06-07 06:45:49,154 INFO     Valid 1p num_queries at step 500: 59078.000000
2021-06-07 06:45:49,155 INFO     Valid average MRR at step 500: 0.777703
2021-06-07 06:45:49,155 INFO     Valid average HITS1 at step 500: 0.679669
2021-06-07 06:45:49,155 INFO     Valid average HITS3 at step 500: 0.860898
2021-06-07 06:45:49,155 INFO     Valid average HITS10 at step 500: 0.931428
...
2021-06-07 06:46:42,893 INFO     Training average positive_sample_loss at step 500: 8.420643
2021-06-07 06:46:42,893 INFO     Training average negative_sample_loss at step 500: 8.420643
2021-06-07 06:46:42,893 INFO     Training average loss at step 500: 8.420643
```

where 500iteration * 2000batch size = 1M samples. This is approximately equivalent to 2 epochs on the original FB15k dataset for knowledge graph completion, which is amazing. Our experience with the negative sampling version of ComplEx is that it doesn't converge so quickly on this dataset, though it's possible the full softmax version of ComplEx can converge faster.

Our questions are
- Is there any insight for the QA iterator? It looks like the distribution of the QA iterator is different from the distribution of the iterator for complex query.
- Are we missing anything important about the training of ComplEx-N3?

**Comment by pminervini** [comment]
Hi @KiddoZhu and @migalkin! Sorry for the massively late answer 🙃 I just got COVID and I'm using the time off to get back to these issues!

I just added the command lines for reproducing the results for each of the types of queries here: https://github.com/pminervini/KGReasoning/blob/main/CQD.md

**Comment by pminervini** [comment]
>  If we understand correctly, the training is carried out with the same codebase, except using only 1p task and the QA iterator.

It's correct!

> We exactly follow the hyperparameters in config.json in the checkpoint directory, and here are the beginning of our training log on FB15k-betae.

I think it's safer if you use the very same command lines we used -- here they are (assuming you refer to the "Query2Box version" of the datasets):

FB15k-237:

```bash
PYTHONPATH=. python3 main.py --do_train --do_valid --do_test --data_path data/FB15k-237-q2b -n 1 -b 2000 -d 1000 -lr 0.1 --warm_up_steps 100000000 --max_steps 100000 --cpu_num 0 --geo cqd --valid_steps 500 --tasks 1p --print_on_screen --test_batch_size 1000 --optimizer adagrad --reg_weight 0.1 --log_steps 500 --cuda --use-qa-iterator
```

FB15k:

```bash
PYTHONPATH=. python3 main.py --do_train --do_valid --do_test --data_path data/FB15k-q2b -n 1 -b 5000 -d 1000 -lr 0.1 --warm_up_steps 100000000 --max_steps 100000 --cpu_num 0 --geo cqd --valid_steps 500 --tasks 1p --print_on_screen --test_batch_size 1000 --optimizer adagrad --reg_weight 0.01 --log_steps 500 --cuda --use-qa-iterator
```

NELL:

```bash
PYTHONPATH=. python3 main.py --do_train --do_valid --do_test --data_path data/NELL-q2b -n 1 -b 2000 -d 1000 -lr 0.1 --warm_up_steps 100000000 --max_steps 100000 --cpu_num 0 --geo cqd --valid_steps 500 --tasks 1p --print_on_screen --test_batch_size 1000 --optimizer adagrad --reg_weight 0.1 --log_steps 500 --cuda --use-qa-iterator
```

Do these let you match the training logs? Could you share the command lines you were using?
It can be that e.g. KGReasoning enables warmup by default, but ComplEx-N3 requires you not to enable that during training -- let me look a bit deeper into this.

**Comment by pminervini** [comment]
I just re-ran these commands and the results match what's in the training logs -- I'm closing this issue, feel free to re-open it if it didn't work for you as well!

After digging a bit, I believe the cause is the missing `--warm_up_steps 100000000` flag to disable warmup -- we disabled it with another option which was not included in this version of the codebase (since `--warm_up_steps` was already there). The learning rate schedule used by other models in `KGReasoning` doesn't seem to work well with ComplEx-N3!
